README.md
Library to safely execute code without fear of the TASK\_UNINTERRUPTIBLE state
-----------------------------------------------------------------------------
thread\_timeout allows to run piece of the python code safely regardless 
of TASK\_UNINTERRUPTIBLE issues.

It provides single decorator, adding a timeout for the function call.


Example of the usage:

    import thread_timeout

    @thread_timeout(10, kill=False)
    def NFS_read(path):
        file(path, 'r').read()

    try:
        print("Result: %s" % NFS_read('broken_nfs/file'))
    except ExecTimeoutException:
        print ("NFS seems to be hung")


thread\_timeout works by running specified function in separate thread and waiting
for timeout (or finalization) of the thread to return value or raise Exception.
If thread is not finished before timeout, thread\_timeout may try to terminate
thread according to kill\_wait value (see below).

thread\_timeout(timeout, kill=True, kill\_wait=0.1)

timeout - seconds, floating, how long to wait thread.
kill - if True (default) attempt to terminate thread with function
kill\_wait - how long to wait after killing before reporting an unresponsive thread 

THREAD KILLING

Thread killing implemented on python level: it will terminate python code, but will not terminate any IO operations or subprocess calls.

Exceptions:

ExecTimeoutException - function did not finish on time, timeout (base class for all following exceptions)
KilledExecTimeoutException - there was a timeout and thread with function was killed successfully
FailedKillExecTimeoutException - there was a timeout and kill attempt but the thread refuses to die
NotKillExecTimeoutException - there was a timeout and there was no attempt to kill thread



debian/changelog
thread-timeout (1.1-1ubuntu1) unstable; urgency=low

  * Add continue_in_fork() function

 -- George Shuklin <george.shuklin@gmail.com>  Wed, 25 Nov 2015 17:03:26 +0200

thread-timeout (1.0-1) unstable; urgency=low

  * Initial debianization

 -- George Shuklin <george.shuklin@gmail.com>  Fri, 09 Oct 2015 18:04:45 +0300


run_test.py
#!/usr/bin/python

import pytest
import time
from thread_timeout import *

def test_no_timeout():
    ''' timeout is not stopping quick function
    '''
    @thread_timeout(2)
    def func(delay):
        time.sleep(delay)

    func(0.5)


def test_raises():
    'ExecTimeout is raised'
    @thread_timeout(1)
    def func(delay):
        time.sleep(delay)

    with pytest.raises(ExecTimeout):
        func(3)

def test_function_return():
    'function returns result'
    @thread_timeout(1)
    def func(x):
        return x

    assert( func('OK') == 'OK')


def test_FailedKillExecTimeout():
    @thread_timeout(1)
    def looong():
        time.sleep(3)
    with pytest.raises(FailedKillExecTimeout):
        looong()


def test_NotKillExecTimeout():
    @thread_timeout(1, kill=False)
    def looong_and_unkillable():
            time.sleep(2)
    try:
        looong_and_unkillable()
        raise Exception('NotKillExecTimeout was expected')
    except NotKillExecTimeout as e:
        print("Test5 OK, got expected exception %s" % repr(e))


def test_KilledExecTimeout():
    @thread_timeout(1, kill_wait=0.40)
    def killme():
        for a in range(0, 200):
            time.sleep(0.01)
        raise Exception("Not killed!")

    try:
        killme()
    except KilledExecTimeout:
        pass


def test_decorator_return():
    ''' decorator is not changing python's into inspection
    '''
    from inspect import getargspec

    def func(x, y=1, *args, **kwargs):
        return vars()

    func_with_timeout = thread_timeout(1)(func)
    assert getargspec(func) == getargspec(func_with_timeout)


def test_class_methods_are_ok():
    class Class(object):

        @thread_timeout(1)
        def short(self, x):
            return x

        @thread_timeout(1, kill_wait=0.33)
        def looong(self, x):
            for x in range(0, 300):
                time.sleep(0.01)
            return x

    obj = Class()
    res = obj.short("OK")
    assert res == 'OK'
    try:
        res = obj.looong('KO')
    except KilledExecTimeout:
        pass


def test_exceptions():
    @thread_timeout(1, kill=False)
    def exception(e):
        raise e

    exception_list = (
                    OverflowError,
                    ReferenceError,
                    SyntaxError,
                    ZeroDivisionError,
                    FloatingPointError,
                    BufferError,
                    LookupError,
                    AssertionError,
                    AttributeError,
                    TypeError,
                    EOFError,
                    IOError,
                    ImportError,
                    IndexError,
                    KeyError,
                    KeyboardInterrupt,
                    MemoryError,
                    NameError,
                    NotImplementedError,
                    OSError,
                    UnboundLocalError,
                    UnicodeError,
                    ValueError,
                    ExecTimeout
                    )
    for exc in exception_list:
        try:
            exception(exc)
        except exc as e:
            pass


def test_decorator_stacking_inner():
    @thread_timeout(2)
    def outer():
        @thread_timeout(1)
        def inner():
            for x in range(0, 500):
                time.sleep(0.01)
        inner()

    begin = time.time()
    try:
        outer()
    except ExecTimeout as e:
        pass
    assert 1 < time.time() - begin < 2


def test_decorator_stacking_outer():
    @thread_timeout(1)
    def outer():
        @thread_timeout(2)
        def inner():
            for x in range(0, 500):
                time.sleep(0.01)
        inner()

    begin = time.time()
    try:
        outer()
    except ExecTimeout:
        pass
    assert 1 < time.time() - begin < 2


def test_waits():
    '''Check if decorator waits before kill'''
    @thread_timeout(3)
    def outer():
        @thread_timeout(3)
        def inner():
            for x in range(0, 100):
                time.sleep(0.01)
        inner()

    begin = time.time()
    outer()
    assert 1 < time.time() - begin < 3

def test_kwarg():
    '''check if we can instance methods with one kwarg (real bug case)'''
    class Foo:
        @thread_timeout(1)
        def bar(self, arg=1):
            pass

    foo = Foo()
    foo.bar(arg=0)

if __name__ == "__main__":
    print("Running tests")

    run(argv=[
        '', __file__,
        '-v'
    ])


fork_test.py
import thread_timeout
import os
import time

print "mypid", os.getpid()
thread_timeout.continue_in_fork()
print "mypid", os.getpid()
time.sleep(15)


thread_timeout/__init__.py
#!/usr/bin/python
# (c) George Shuklin, 2015
#
# This library is free software; you can redistribute it and/or
# Modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# Version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# But WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
from __future__ import print_function
import threading
import time
import sys
import ctypes
import wrapt  # pip install wrapt, apt-get install python-wrapt
from Queue import Queue
import signal
import os
import atexit
'''
    thread_timeout decorator allows to run piece of the python code
    safely regardless of TASK_UNINTERRUPTIBLE issues ('D' state).

    Main sources of 'D' state are broken NFS, bad disk/IO, or kernel bugs.

    Library provides single decorator, adding a timeout for the function call.


    Example of the usage:
        import thread_timeout

        @thread_timeout(10, kill=False)
        def NFS_read(path):
            file(path, 'r').read()

        try:
            print("Result: %s" % NFS_read('/broken_nfs/file'))
        except ExecTimeout:
            print ("NFS seems to be hung")


    thread_timeout works by running specified function in separate
    thread and waiting for timeout (or finalization) of the thread
    to return value or raise exception.
    If thread is not finished before timeout, thread_timeout will
    try to terminate thread according to kill value (see below).

    thread_timeout(timeout, kill=True, kill_wait=0.1)

    timeout - seconds, floating, how long to wait thread.
    kill - if True (default) attempt to terminate thread with function
    kill_wait - how long to wait after killing before reporting
    an unresponsive thread

    THREAD KILLING

    Thread killing implemented on python level: it will terminate python
    code, but will not terminate any IO operations or subprocess calls.

    Exceptions:

    ExecTimeout - function did not finish on time, timeout
        (base class for all following exceptions)
    KilledExecTimeout - there was a timeout and thread
        with function was killed successfully
    FailedKillExecTimeout - there was a timeout and kill attempt
        but the thread refuses to die
    NotKillExecTimeout - there was a timeout and there
        was no attempt to kill thread
'''


def _kill_thread(thread):
    # heavily based on http://stackoverflow.com/a/15274929/2281274
    # by Johan Dahlin
    # rewrited to avoid licence uncertainty

    # due to the strangeness in python 2.x, thread killing happens
    # within 32 python operations regardless of duration
    # (f.e. 32 x sleep(1), or 32 x sleep (0.01))
    # python3 works fine
    import sys
    SE = ctypes.py_object(SystemExit)
    tr = ctypes.c_long(thread.ident)
    res = ctypes.pythonapi.PyThreadState_SetAsyncExc(tr, SE)


class ExecTimeout(BaseException):
    pass


class KilledExecTimeout(ExecTimeout):
    pass


class FailedKillExecTimeout(ExecTimeout):
    pass


class NotKillExecTimeout(ExecTimeout):
    pass

class ReturnValue:
    type = None
    value = None
    exception = None
    

def thread_timeout(delay, kill=True, kill_wait=0.04):
    @wrapt.decorator
    def wrapper(wrapped, instance, args, kwargs):
        queue = Queue()

        def inner_worker():
            try:
                result = wrapped(*args, **kwargs)
                ret_value = ReturnValue()
                ret_value.type = 'success'
                ret_value.value = result
                queue.put(ret_value)
            except:
                e = sys.exc_info()
                ret_value = ReturnValue()
                ret_value.type = 'exception'
                ret_value.exception = e
                queue.put(ret_value)
        thread = threading.Thread(target=inner_worker)
        thread.daemon = True
        thread.start()
        thread.join(delay)
        if thread.isAlive():
            if not kill:
                raise NotKillExecTimeout(
                    "Timeout and no kill attempt")
            _kill_thread(thread)
            time.sleep(kill_wait)
            # FIXME isAlive is giving fals positive results
            if thread.isAlive():
                raise FailedKillExecTimeout(
                    "Timeout, thread refuses to die in %s seconds" %
                    kill_wait)
            else:
                raise KilledExecTimeout(
                    "Timeout and thread was killed")
        res = queue.get()
        if res.type == 'success':
            return res.value
        if res.type == 'exception':
            raise res.exception[0], res.exception[1], res.exception[2]
    return wrapper

def continue_in_fork():
    '''
        Run rest of application as separate process
        current process goes to endless sleep and passes
        all (possible) signals back to child.
        (child PID stored in CHILD_PID variable in the module)
        Obviously terminates itself on SIGKILL.

        child register an exit handler to terminate parent.

        To avoid complications, it uses clib's fork function via ctypes.
    '''
    CHILD_PID = 0
    PARENT_PID = 0
    def all_signals_handler(sign, frame):
        if CHILD_PID:
            os.kill(CHILD_PID, sign)
        if sign == signal.SIGTERM:
            raise SystemExit

    def terminate_parent():
        if PARENT_PID:
            os.kill(PARENT_PID, signal.SIGTERM)

    PARENT_PID = os.getpid()
    libc = ctypes.CDLL("libc.so.6")
    pid = libc.fork()
    if pid == -1:
        raise RuntimeError("Unable to fork")
    if pid == 0:  # CHILD
        atexit.register(terminate_parent)
        return   # continue to run app in a new process
    else:  # PARENT THREAD
        CHILD_PID = pid
        uncatchable = ['SIG_DFL','SIGSTOP','SIGKILL']
        for i in [x for x in dir(signal) if x.startswith("SIG")]:
            if not i in uncatchable:
                signum = getattr(signal,i)
                signal.signal(signum,all_signals_handler)
        while True:
            time.sleep(31536000)  # 1 year
